[REST] Handle UTXO retrieval when ignoring the mempool

romanz opened the pull request:
Current REST API always returns empty UTXO when invoked without `/checkmempool/` URL part.

After the fix:
```
$ curl -s http://localhost:8332/rest/getutxos/0e3e2357e806b6cdb1f70b54c3a3a17b6714ee1f0e68bebb44a74b1efd512098-0.json | jq
{
  "chainHeight": 514109,
  "chaintipHash": "0000000000000000001fe76d1445e8a6432fd2de04261dc9c5915311dc7ad6de",
  "bitmap": "1",
  "utxos": [
    {
      "height": 1,
      "value": 50,
      "scriptPubKey": {
        "asm": "0496b538e853519c726a2c91e61ec11600ae1390813a627c66fb8be7947be63c52da7589379515d4e0a604f8141781e62294721166bf621e73a82cbf2342c858ee OP_CHECKSIG",
        "hex": "410496b538e853519c726a2c91e61ec11600ae1390813a627c66fb8be7947be63c52da7589379515d4e0a604f8141781e62294721166bf621e73a82cbf2342c858eeac",
        "reqSigs": 1,
        "type": "pubkey",
        "addresses": [
          "12c6DSiU4Rq3P4ZxziKxzrL5LmMBrzjrJX"
        ]
      }
    }
  ]
}
```

Before the fix:
```
$ curl -s http://localhost:8332/rest/getutxos/0e3e2357e806b6cdb1f70b54c3a3a17b6714ee1f0e68bebb44a74b1efd512098-0.json | jq
{
  "chainHeight": 514109,
  "chaintipHash": "0000000000000000001fe76d1445e8a6432fd2de04261dc9c5915311dc7ad6de",
  "bitmap": "0",
  "utxos": []
}
```

romanz committed:
[REST] Handle UTXO retrieval when ignoring the mempool

Current REST API always returns empty UTXO when invoked without `/checkmempool/` URL part.

After the fix:
```
$ curl -s http://localhost:8332/rest/getutxos/0e3e2357e806b6cdb1f70b54c3a3a17b6714ee1f0e68bebb44a74b1efd512098-0.json | jq
{
  "chainHeight": 514109,
  "chaintipHash": "0000000000000000001fe76d1445e8a6432fd2de04261dc9c5915311dc7ad6de",
  "bitmap": "1",
  "utxos": [
    {
      "height": 1,
      "value": 50,
      "scriptPubKey": {
        "asm": "0496b538e853519c726a2c91e61ec11600ae1390813a627c66fb8be7947be63c52da7589379515d4e0a604f8141781e62294721166bf621e73a82cbf2342c858ee OP_CHECKSIG",
        "hex": "410496b538e853519c726a2c91e61ec11600ae1390813a627c66fb8be7947be63c52da7589379515d4e0a604f8141781e62294721166bf621e73a82cbf2342c858eeac",
        "reqSigs": 1,
        "type": "pubkey",
        "addresses": [
          "12c6DSiU4Rq3P4ZxziKxzrL5LmMBrzjrJX"
        ]
      }
    }
  ]
}
```

Before the fix:
```
$ curl -s http://localhost:8332/rest/getutxos/0e3e2357e806b6cdb1f70b54c3a3a17b6714ee1f0e68bebb44a74b1efd512098-0.json | jq
{
  "chainHeight": 514109,
  "chaintipHash": "0000000000000000001fe76d1445e8a6432fd2de04261dc9c5915311dc7ad6de",
  "bitmap": "0",
  "utxos": []
}
```

ryanofsky reviewed (commented):
utACK b1038afceadacb640ab1fa4fd9b577bb51d45e84

promag commented:
utACK b1038af.

Nit, could get rid of `viewDummy` by using `viewChain` by default:
```cpp
        LOCK2(cs_main, mempool.cs);

        CCoinsViewCache& viewChain = *pcoinsTip;
        CCoinsViewMemPool viewMempool(&viewChain, mempool);
        CCoinsViewCache view(&viewChain);
        if (fCheckMemPool) {
            // switch cache backend to db+mempool in case user likes to query mempool
            view.SetBackend(viewMempool);
        }
```

romanz commented:
Thanks for the review!
I removed `viewDummy` as suggested.

promag commented:
@eklitzke that gives:
```cpp
rest.cpp:495:44: error: incompatible operand types ('CCoinsViewMemPool *' and 'CCoinsViewCache *')
```
it needs a cast:
```cpp
CCoinsViewCache view(fCheckMemPool ? (CCoinsView*) &viewMempool : &viewChain);
```
But yeah, either way.

promag commented:
But I wonder if it pays off dealing with both cases separately:
```cpp
if (fCheckMemPool) {
    LOCK2(cs_main, mempool.cs);
    CCoinsViewCache& viewChain = *pcoinsTip;
    CCoinsViewMemPool viewMempool(&viewChain, mempool);
    CCoinsViewCache view(&viewMempool);
    ...
} else {
    LOCK(cs_main);  // no need to lock mempool!
    CCoinsViewCache& viewChain = *pcoinsTip;
    CCoinsViewCache view(&viewChain);
    ...
}
```

romanz commented:
@eklitzke @promag I'm fine with both suggestions, please let me know which is the preferred one :)

eklitzke commented:
The way @promag just suggested is best of all. If you want to do that you'll have to refactor the loop out so it can be called in both code paths. You can do that by adding a new static method or using a lambda.

romanz committed:
Make CTxMemPool::isSpent() const

romanz commented:
Thanks!
I have updated the code according (refactoring the loop into a lambda), and added a few more tests (to make sure that we don't return outputs, being spent by mempool transactions).

jnewbery commented on test/functional/interface_rest.py:
```diff
@@ -151,23 +151,48 @@ def run_test(self):
         json_string = http_get_call(url.hostname, url.port, '/rest/getutxos'+json_request+self.FORMAT_SEPARATOR+'json')
         json_obj = json.loads(json_string)
-        assert_equal(len(json_obj['utxos']), 0) #there should be an outpoint because it has just added to the mempool
+        assert_equal(len(json_obj['utxos']), 0) #there should be no outpoint because it has just added to the mempool
```
I think this comment could be clearer. Something like `# tx not found since it is in the mempool (and /mempool not included in request`

edit: in fact, I think you could remove all the inline comments and just place a comment at the top of the section explaining what's happening (create a transaction, check that it's found with `/checkmempool`, but not found without, then confirm the transaction and check that it's found with or without `/checkmempool`

romanz replied:
Thanks!
I agree, and would be happy to fix this issue (together with a general tidying-up of `interface_rest.py` tests) in a following PR.

jnewbery reviewed (commented):
Tested ACK eb0aad1c0b90893caadd1c96d93b3f2572301cf3 with one nit in the test, which you can take or leave.

`interface_rest.py` could do with some general tidy-up, so it may be best to just leave that for a future PR.

romanz commented:
Thanks!
I agree, and would be happy to fix this issue (together with a general tidying-up of `interface_rest.py` tests) in a following PR.

sipa commented on src/rest.cpp:
```diff
@@ -487,26 +487,33 @@ static bool rest_getutxos(HTTPRequest* req, const std::string& strURIPart)
+            LOCK2(cs_main, mempool.cs);
+            CCoinsViewCache& viewChain = *pcoinsTip;
+            CCoinsViewMemPool viewMempool(&viewChain, mempool);
+            CCoinsViewCache view(&viewMempool);
```
There's no need for an intermediary cache here anymore. You can make the lambda take a `CCoinsView&` instead, and either pass it `viewMempool` (here) or `*pcoinsTip` (below) directly.

romanz replied:
You're right, thanks for the catch!
Will amend the commit, and update this PR.

jnewbery commented:
> I agree, and would be happy to fix this issue (together with a general tidying-up of interface_rest.py tests) in a following PR.

Great! I already started tidying up the test here: https://github.com/jnewbery/bitcoin/tree/tidy_up_rest_test . I haven't opened a PR yet. Let me know what you think (and feel free to take the branch, modify as you see fit and open a PR if you want to take this on).

sipa commented:
utACK 5e2cfcba4a46a6d39f38f03dcd30543d259f181b

romanz commented:
@jnewbery The branch looks great :)
I will continue the development at https://github.com/romanz/bitcoin/tree/tidy_up_rest_test in the upcoming days.
Much appreciated, and thanks again!

promag commented on src/rest.cpp:
```diff
@@ -487,26 +487,31 @@ static bool rest_getutxos(HTTPRequest* req, const std::string& strURIPart)
+            for (const COutPoint& vOutPoint : vOutPoints) {
+                bool hit = false;
+                Coin coin;
+                if (view.GetCoin(vOutPoint, coin) && !mempool.isSpent(vOutPoint)) {
```
Alternative:
```cpp
Coin coin;
const bool hit = !mempool.isSpent(vOutPoint) && view.GetCoin(vOutPoint, coin);
hits.push_back(hit);
if (hit) outs.emplace_back(std::move(coin));
```
Also, cheapest condition to fail should come first, and I think `!isSpent` is the cheapest in both cases.

romanz replied:
Thanks for the suggestion!
I have updated and rebased the PR.

romanz commented:
>  In general it's better to avoid rebase if possible since it makes re-reviewing easier if old and new branches are from the same base commit.

Will do, thanks for the tip!
